File: ios/EatPal/EatPal/Models/Recipe.swift
import Foundation

struct NutritionInfo: Codable, Equatable {
    var calories: Double?
    var proteinG: Double?
    var carbsG: Double?
    var fatG: Double?
    var fiberG: Double?
    var calciumMg: Double?
    var ironMg: Double?

    enum CodingKeys: String, CodingKey {
        case calories
        case proteinG = "protein_g"
        case carbsG = "carbs_g"
        case fatG = "fat_g"
        case fiberG = "fiber_g"
        case calciumMg = "calcium_mg"
        case ironMg = "iron_mg"
    }
}

struct Recipe: Identifiable, Codable, Equatable {
    let id: String
    var userId: String
    var householdId: String?
    var name: String
    var description: String?
    var instructions: String?
    var foodIds: [String]
    var category: String?
    var prepTime: String?
    var cookTime: String?
    var totalTimeMinutes: Int?
    var servings: String?
    var servingsMin: Int?
    var servingsMax: Int?
    var defaultServings: Int?
    var additionalIngredients: String?
    var tips: String?
    var imageUrl: String?
    var sourceUrl: String?
    var sourceType: String?
    var tags: [String]?
    var rating: Double?
    var timesMade: Int?
    var lastMadeDate: String?
    var difficultyLevel: String?
    var kidFriendlyScore: Double?
    var nutritionInfo: NutritionInfo?
    var createdAt: String?
    var updatedAt: String?

    enum CodingKeys: String, CodingKey {
        case id
        case userId = "user_id"
        case householdId = "household_id"
        case name, description, instructions, category, tags, rating, tips
        case foodIds = "food_ids"
        case prepTime = "prep_time"
        case cookTime = "cook_time"
        case totalTimeMinutes = "total_time_minutes"
        case servings
        case servingsMin = "servings_min"
        case servingsMax = "servings_max"
        case defaultServings = "default_servings"
        case additionalIngredients = "additional_ingredients"
        case imageUrl = "image_url"
        case sourceUrl = "source_url"
        case sourceType = "source_type"
        case timesMade = "times_made"
        case lastMadeDate = "last_made_date"
        case difficultyLevel = "difficulty_level"
        case kidFriendlyScore = "kid_friendly_score"
        case nutritionInfo = "nutrition_info"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }

    mutating func apply(_ updates: RecipeUpdate) {
        if let name = updates.name { self.name = name }
        if let description = updates.description { self.description = description }
        if let instructions = updates.instructions { self.instructions = instructions }
        if let foodIds = updates.foodIds { self.foodIds = foodIds }
        if let category = updates.category { self.category = category }
        if let prepTime = updates.prepTime { self.prepTime = prepTime }
        if let cookTime = updates.cookTime { self.cookTime = cookTime }
        if let servings = updates.servings { self.servings = servings }
        if let tags = updates.tags { self.tags = tags }
        if let rating = updates.rating { self.rating = rating }
        if let difficultyLevel = updates.difficultyLevel { self.difficultyLevel = difficultyLevel }
    }
}

struct RecipeUpdate: Codable {
    var name: String?
    var description: String?
    var instructions: String?
    var foodIds: [String]?
    var category: String?
    var prepTime: String?
    var cookTime: String?
    var servings: String?
    var tags: [String]?
    var rating: Double?
    var difficultyLevel: String?

    enum CodingKeys: String, CodingKey {
        case name, description, instructions, category, tags, rating, servings
        case foodIds = "food_ids"
        case prepTime = "prep_time"
        case cookTime = "cook_time"
        case difficultyLevel = "difficulty_level"
    }
}

File: android-native/app/src/main/java/com/eatpal/app/models/Recipe.kt
package com.eatpal.app.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class NutritionInfo(
    val calories: Double? = null,
    @SerialName("protein_g") val proteinG: Double? = null,
    @SerialName("carbs_g") val carbsG: Double? = null,
    @SerialName("fat_g") val fatG: Double? = null,
    @SerialName("fiber_g") val fiberG: Double? = null,
    @SerialName("calcium_mg") val calciumMg: Double? = null,
    @SerialName("iron_mg") val ironMg: Double? = null,
)

/** Mirrors iOS `Recipe`. */
@Serializable
data class Recipe(
    val id: String,
    @SerialName("user_id") val userId: String,
    @SerialName("household_id") val householdId: String? = null,
    val name: String,
    val description: String? = null,
    val instructions: String? = null,
    @SerialName("food_ids") val foodIds: List<String> = emptyList(),
    val category: String? = null,
    @SerialName("prep_time") val prepTime: String? = null,
    @SerialName("cook_time") val cookTime: String? = null,
    @SerialName("total_time_minutes") val totalTimeMinutes: Int? = null,
    val servings: String? = null,
    @SerialName("servings_min") val servingsMin: Int? = null,
    @SerialName("servings_max") val servingsMax: Int? = null,
    @SerialName("default_servings") val defaultServings: Int? = null,
    @SerialName("additional_ingredients") val additionalIngredients: String? = null,
    val tips: String? = null,
    @SerialName("image_url") val imageUrl: String? = null,
    @SerialName("source_url") val sourceUrl: String? = null,
    @SerialName("source_type") val sourceType: String? = null,
    val tags: List<String>? = null,
    val rating: Double? = null,
    @SerialName("times_made") val timesMade: Int? = null,
    @SerialName("last_made_date") val lastMadeDate: String? = null,
    @SerialName("difficulty_level") val difficultyLevel: String? = null,
    @SerialName("kid_friendly_score") val kidFriendlyScore: Double? = null,
    @SerialName("nutrition_info") val nutritionInfo: NutritionInfo? = null,
    @SerialName("created_at") val createdAt: String? = null,
    @SerialName("updated_at") val updatedAt: String? = null,
) {
    fun applying(u: RecipeUpdate): Recipe = copy(
        name = u.name ?: name,
        description = u.description ?: description,
        instructions = u.instructions ?: instructions,
        foodIds = u.foodIds ?: foodIds,
        category = u.category ?: category,
        prepTime = u.prepTime ?: prepTime,
        cookTime = u.cookTime ?: cookTime,
        servings = u.servings ?: servings,
        tags = u.tags ?: tags,
        rating = u.rating ?: rating,
        difficultyLevel = u.difficultyLevel ?: difficultyLevel,
        imageUrl = u.imageUrl ?: imageUrl,
        additionalIngredients = u.additionalIngredients ?: additionalIngredients,
        tips = u.tips ?: tips,
    )
}

@Serializable
data class RecipeUpdate(
    val name: String? = null,
    val description: String? = null,
    val instructions: String? = null,
    @SerialName("food_ids") val foodIds: List<String>? = null,
    val category: String? = null,
    @SerialName("prep_time") val prepTime: String? = null,
    @SerialName("cook_time") val cookTime: String? = null,
    val servings: String? = null,
    val tags: List<String>? = null,
    val rating: Double? = null,
    @SerialName("difficulty_level") val difficultyLevel: String? = null,
    @SerialName("image_url") val imageUrl: String? = null,
    @SerialName("additional_ingredients") val additionalIngredients: String? = null,
    val tips: String? = null,
)
